src/VASSAL/build/module/Chatter.java

```diff
@@ -21,477 +21,636 @@
+				} else {
+					style = "msg";
+					html_allowed = !htmlCompatibility; //BR// Generic report lines check compatibility flag (so old modules will not break on e.g. "<" in messages)
+				}
+			} else if (s.startsWith("-")) {
+				style = "sys";
+				html_allowed = true;
+			} else {
+			  style = getChatStyle(s);
+			  html_allowed = false;
+			}
+		} else {
+			style = "msg";
+			html_allowed = false;
+		}
+
+		//BR// Disable unwanted HTML tags in contexts where it shouldn't be allowed:
+		//BR// (1) Anything received from chat channel, for security reasons
+		//BR// (2) Legacy module "report" text when not explicitly opted in w/ first character or preference setting 
+		if (!html_allowed) {
+		  s = s.replaceAll("<", "&lt;"); //BR// This prevents any unwanted tag from functioning
+      s = s.replaceAll(">", "&gt;"); //BR// This makes sure > doesn't break any of our legit <div> tags
+		}
+
+		//BR// Systematically search through for html image tags. When we find one, try
+		//BR// to match it with an image from our DataArchive, and substitute the correct
+		//BR// fully qualified URL into the tag.
+		
+		//BR// There are without a doubt "more efficient ways to do this", but this loop
+		//BR// miraculously ran perfectly the very first time I compiled it, so superstition...		
+		URL url;
+		String keystring = "<img src=\"";
+		String file, tag, replace;
+		int base;
+		while (s.toLowerCase().contains(keystring)) { // Find next key (to-lower so we're not case sensitive)
+			base = s.toLowerCase().indexOf(keystring);
+			file = s.substring(base + keystring.length(), s.length()).split("\"")[0]; // Pull the filename out from between the quotes
+    	    tag  = s.substring(base, base + keystring.length()) + file + "\"";    // Reconstruct the part of the tag we want to remove, leaving all attributes after the filename alone, and properly matching the upper/lowercase of the keystring						
+			
+			try {
+				url = GameModule.getGameModule().getDataArchive().getURL("images/" + file);
+				replace = "<img  src=\"" + url.toString() + "\""; // Fully qualified URL if we are successful. The extra
+                                                          // space between IMG and SRC in the processed
+                                                          // version ensures we don't re-find THIS tag as we iterate
+			} catch (IOException ex) {
+				replace = "<img  src=\"" + file + "\""; // Or just leave in except alter just enough that we won't find this tag again
+			}
+
+			if (s.contains(tag)) {
+				s = s.replaceFirst(tag, replace); // Swap in our new URL-laden tag for the old one.
+			} else {
+				break; // BR// If something went wrong in matching up the tag, don't loop forever
+			}
+		}
+		
+		//BR// Now we have to fix up any legacy angle brackets around the word <observer>
+    keystring = Resources.getString("PlayerRoster.observer");
+    replace   = keystring.replace("<","&lt;");
+    replace   = replace.replace(">","&gt;");
```

Comment:
Can chain the calls here: `replace = keystring.replace(a, b).replace(c, d);`

Reply:
fixed for next push